### PR TITLE
Add support for Orbit token bridge

### PR DESCRIPTION
**Pull request by gvladika**
This commit adds support for token bridge for Orbit (ERC20-based) rollups where custom token is used to pay for fees.

Implementation is done mainly in new Orbit-specific contracts which inherit from their "standard" counterparts. 
Ie.:
- L1OrbitERC20Gateway inherits from L1ERC20Gateway
- L1OrbitCustomGateway inherits from L1CustomGateway
- L1OrbitGatewayRouter inherits from L1GatewayRouter
- L1OrbitReverseCustomGateway from L1ReverseCustomGateway

Bridging of rollup's native fee token through token bridge is not allowed. It can only be bridged through inbox directly, and in that case it is minted on L2 side as native currency.

Also Includes:
- Foundry unit and integration test cases
- e2e Hardhat test cases executed in local test env

### CODE REVIEW

File: contracts/tokenbridge/ethereum/gateway/L1OrbitReverseCustomGateway.sol [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * @title   L1 Gateway for reverse "custom" bridging functionality in an ERC20-based rollup.
+ * @notice  Handles some (but not all!) reverse custom Gateway needs.
+ *          Use the reverse custom gateway instead of the normal custom
+ *          gateway if you want total supply to be tracked on the L2
+ *          rather than the L1.
+ * @dev     The reverse custom gateway burns on the l2 and escrows on the l1
+ *          which is the opposite of the way the normal custom gateway works
+ *          This means that the total supply L2 isn't affected by briding, which
```
Comment (yahgwai) [inline contracts/tokenbridge/ethereum/gateway/L1OrbitReverseCustomGateway.sol:16]:
```suggestion
 *          This means that the total supply L2 isn't affected by bridging, which
```

File: contracts/tokenbridge/ethereum/gateway/L1OrbitERC20Gateway.sol [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    function outboundTransferCustomRefund(
+        address _l1Token,
+        address _refundTo,
+        address _to,
+        uint256 _amount,
+        uint256 _maxGas,
+        uint256 _gasPriceBid,
+        bytes calldata _data
+    ) public payable override returns (bytes memory res) {
+        require(_l1Token != _getNativeFeeToken(), "NOT_ALLOWED_TO_BRIDGE_FEE_TOKEN");
```
Comment (yahgwai) [inline contracts/tokenbridge/ethereum/gateway/L1OrbitERC20Gateway.sol:28]: I think a comment here about why this isnt allowed would be good

File: contracts/tokenbridge/ethereum/gateway/L1ArbitrumGateway.sol [hunk 3]
```diff
@@ -338,4 +341,65 @@ abstract contract L1ArbitrumGateway is
...
+
+    /**
+     * @notice Intermediate internal function that passes on parameters needed to trigger creation of retryable ticket.
+     * @param _refundTo Account, or its L2 alias if it have code in L1, to be credited with excess gas refund in L2
+     * @param _from Initiator of deposit
+     * @param _amount Token amount being deposited
+     * @param _maxGas Max gas deducted from user's L2 balance to cover L2 execution
+     * @param _gasPriceBid Gas price for L2 execution
+     * @param _maxSubmissionCost Max gas deducted from user's L2 balance to cover base submission fee
+     * //param tokenTotalFeeAmount Amount of fees to be deposited in native token to cover for retryable ticket cost (used only in ERC20-based rollups)
```
[the remark this answers is not in the document]
Comment (gvladika) [reply contracts/tokenbridge/ethereum/gateway/L1ArbitrumGateway.sol]: done

File: contracts/tokenbridge/ethereum/gateway/L1OrbitERC20Gateway.sol (hunk 2 above)
Comment (gvladika) [reply contracts/tokenbridge/ethereum/gateway/L1OrbitERC20Gateway.sol:30]: done

**Pull request merged**